### PR TITLE
items: use temporary item_type for circulation

### PR DESCRIPTION
**Pull request by zannkukai**
If a temporary item_type is defined and valid on an item, it must be
used for circulation operation (checkout, extend, renew).

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>
Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/items/api/record.py [hunk 1]
```diff
@@ -361,6 +361,26 @@ def item_type_pid(self):
...
+            if end_date:
+                now_date = pytz.utc.localize(datetime.now())
+                end_date = date_string_to_utc(end_date)
+                if now_date > end_date:
+                    return None
+            return extracted_data_from_ref(tmp_item_type.get('$ref'))
+
+    @property
+    def item_type_circulation_category_pid(self):
+        """Shortcut to find the best item_type too use for circulation."""
```
Comment (iGormilhit) [inline rero_ils/modules/items/api/record.py:381]:
```suggestion
        """Shortcut to find the best item_type to use for circulation."""
```

File: rero_ils/modules/items/tasks.py [hunk 2]
```diff
@@ -64,3 +66,33 @@ def process_late_claimed_issues(
...
     return msg
+
+
+@shared_task
+def clean_obsolete_temporary_item_types():
+    """Clean obsoletes temporary item_type for items.
+
+    Search for all item with obsolete temporary item_type. For each found item
+    clean the temporary item_type informations. Update item into database to
+    commit change
```
Comment (iGormilhit) [inline rero_ils/modules/items/tasks.py:77]:
```suggestion
    Search for all item with obsolete temporary item_type. For each found item,
    clean the temporary item_type informations. Update item into database to
    commit change
```

File: tests/api/circulation/test_temporary_item_type.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline tests/api/circulation/test_temporary_item_type.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message proposition:

```
items: use temporary item_type for circulation

* Uses a temporary item_type (or circulation category) for circulation
  operations (checkout, extend, renew) if it's defined and valid on an
  item.
* Creates a new cron tasks to clear the obsolete temporary item_type.
```
- No dot at the end of the commit title.
- If a sentence start with a verb (3rd singular person: *creates*), then a list is better.

**Review by iGormilhit — APPROVED** [review]

**Comment by benerken** [comment]
- when you add a temporary circulation category WITHOUT indicating a "end date", the system doesn't display the temporary circulation in the document detail/item list.

**Review by benerken — APPROVED** [review]

File: rero_ils/modules/items/api/api.py [hunk 4]
```diff
@@ -165,6 +174,27 @@ def get_item_record_for_ui(cls, **kwargs):
...
+
+        An end_date could be attached to the item temporary item_type. If this
+        date is less or equal to sysdate, then the temporary_item_type must be
+        considered as obsolete.
+
+        :param end_date: the end_date to check (`datetime.now()` by default)
+        :return A generator of `ItemRecord` object.
+        """
+        if end_date is None:
+            end_date = datetime.now()
```
Comment (rerowep) [inline rero_ils/modules/items/api/api.py:189]: Do we have to use `utcnow`?

File: rero_ils/modules/items/tasks.py [hunk 5]
```diff
@@ -64,3 +66,33 @@ def process_late_claimed_issues(
...
+                item_pid=item.pid,
+                tmp_itty=tmp_itty.get('name'),
+                tmp_date=tmp_itty_enddate,
+                default_itty=default_itty.get('name')
+            )
+        )
+
+        # remove the obsolete data
+        del item['temporary_item_type']
+        item.update(data=item, dbcommit=True, reindex=True)
```
Comment (rerowep) [inline rero_ils/modules/items/tasks.py:98]: Maybe we have to use `item.replace` here?

Comment (zannkukai) [reply rero_ils/modules/items/tasks.py:98]: I don't know this function. I will check it, it seems interesting

**Review by BadrAly — APPROVED** [review]

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

**Pull request merged**